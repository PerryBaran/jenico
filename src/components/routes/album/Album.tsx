import { Dispatch, SetStateAction } from "react";
import { useParams } from "react-router-dom";
import { SongInfo, Songs } from "../../../Interface";
import style from "./album.module.css";
import { playCircle, pauseCircle } from "../../../media/icons/index";

interface Props {
  data: SongInfo[];
  playing: boolean;
  setPlaying: Dispatch<SetStateAction<boolean>>;
  selectedSong: string;
  setSelectedSong: Dispatch<SetStateAction<string>>;
}

function Album(props: Props) {
  const { data, playing, setPlaying, selectedSong, setSelectedSong } = props;
  const params = useParams();
  const album = data.find((album) => album.title === params.album);

  const handlePlayAlbum = () => {
    if (!album) return;
    const { songs } = album;
    if (songs.some((song) => song.name === selectedSong)) {
      setPlaying((prev) => !prev);
    } else {
      setSelectedSong(songs[0].name);
      setPlaying(true);
    }
  };

  const handlePlaySong = (name: string) => {
    setSelectedSong(name);
    setPlaying(true);
  };

  return (
    <>
      <div
        className={style.albumBackground}
        style={{ backgroundImage: `url(${album?.art})` }}
      />
      <div className={style.albumContainer}>
        <div>
          <h2>
            <a href={album?.hyperlink} target="_blank" rel="noreferrer">
              {album?.title}
            </a>
          </h2>
          <div className={style.album}>
            <button onClick={handlePlayAlbum}>
              <img src={album?.art} alt={`Cover for ${album?.title}`} />
              <img
                src={
                  playing &&
                  album?.songs.find((song) => song.name === selectedSong)
                    ? pauseCircle
                    : playCircle
                }
                alt="play/pause album"
              />
            </button>
            {album?.songs.length && album.songs.length > 1 && (
              <ul>
                {album?.songs.map((song: Songs) => {
                  return (
                    <li
                      key={song.name}
                      className={
                        song.name === selectedSong ? style.playing : undefined
                      }
                    >
                      <button onClick={() => handlePlaySong(song.name)}>
                        {song.name}
                      </button>
                    </li>
                  );
                })}
              </ul>
            )}
          </div>
        </div>
      </div>
    </>
  );
}

export default Album;
