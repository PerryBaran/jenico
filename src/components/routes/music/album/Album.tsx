import { useEffect, useState, Dispatch, SetStateAction } from 'react';
import { useParams } from 'react-router-dom';
import { SongInfo, Songs } from '../../../../Interface';
import style from './album.module.css';
import { playCircle, pauseCircle } from '../../../../media/icons/index';

interface Props {
    data: SongInfo[], 
    playing: boolean, 
    setPlaying: Dispatch<SetStateAction<boolean>>, 
    songIndex: number,
    setSongIndex: Dispatch<SetStateAction<number>>, 
    albumIndex: number, 
    setAlbumIndex: Dispatch<SetStateAction<number>>
}

function Album(props: Props) {
    const {data, playing, setPlaying, songIndex, setSongIndex, albumIndex, setAlbumIndex} = props;
    const params = useParams();
    const [info, setInfo] = useState<null|SongInfo>(null);
    const [pageIndex, setPageIndex] = useState(0);

    useEffect(() => {
        for (let i = data.length - 1; i >= 0; i--) {
            if (data[i].title === params.id) {
                setInfo(data[i]);
                setPageIndex(i);
            };
        };
    //eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const renderSongList = () => {
        if (info?.songs.length && info.songs.length > 1) {
            return (
                <ul className={style.songList}>
                    {info?.songs.map((song: Songs, i: number) => {
                        return (
                            <li key={song.name} className={`${isSongPlaying(i)} ${style.song}`}>
                                <button onClick={() => playSelectedSong(i)}>{song.name}</button>
                            </li>
                        )
                    })}
                </ul>
            )
        };
    };

    const isPageAlbumPlaying = () => {
        if (playing && pageIndex === albumIndex) {
            return pauseCircle
        }
        return playCircle
    };

    const isSongPlaying = (index: number) => {
        if (pageIndex === albumIndex && songIndex === index) {
            return style.playing
        }
        return ''
    };

    const playCurrentAlbum = () => {
        if (pageIndex === albumIndex) {
            setPlaying(!playing)
        } else {
            setAlbumIndex(pageIndex);
            setSongIndex(0);
            setPlaying(true);
        }
    };

    const playSelectedSong = (index: number) => {
        if (pageIndex === albumIndex) {
            setSongIndex(index);
            setPlaying(true);
        } else {
            setAlbumIndex(pageIndex);
            setSongIndex(index);
            setPlaying(true);
        }
    };

    return (
        <div>
            <div className={style.albumBackground} style={{backgroundImage: `url(${info?.art})`}}/>
            <div className={style.album}>
                <div>
                    <h2><a href={info?.hyperlink} target='_blank' rel="noreferrer">{info?.title}</a></h2>
                    <div className={style.cover}>
                        <button className={style.playAlbum} onClick={() => playCurrentAlbum()}>
                            <img src={info?.art} alt={`Cover for ${info?.title}`}/>
                            <img src={isPageAlbumPlaying()} alt='play/pause album' className={style.playPause}/>
                        </button>
                        {renderSongList()}
                    </div> 
                </div>
            </div>
        </div>
    );
};

export default Album;