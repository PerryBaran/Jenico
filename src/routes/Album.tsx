import React from 'react';

function Album(props:any) {
    return (
        <div>
            
        </div>
    );
}

export default Album;