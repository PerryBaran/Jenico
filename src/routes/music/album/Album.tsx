import React, { useEffect, useState } from 'react';
import {useParams} from 'react-router-dom';
import { SongInfo, Songs } from '../../../Interface';
import style from './album.module.css';

function Album(props: {data: SongInfo[]}) {
    const {data} = props
    const params = useParams()
    const [info, setInfo] = useState<null|SongInfo>(null)

    useEffect(() => {
        for (let i = data.length - 1; i >= 0; i--) {
            if (data[i].title === params.id) {
                setInfo(data[i])
            }
        }
    //eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const renderSongList = () => {
        if (info?.songs.length && info.songs.length > 1) {
            return (
                <ul className={style.songList}>
                    {info?.songs.map((song: Songs) => {
                        return (
                            <li key={song.name}>
                                <button>{song.name}</button>
                            </li>
                        )
                    })}
                </ul>
            )
        }
    }

    return (
        <div>
            <div className={style.albumBackground} style={{backgroundImage: `url(${info?.art})`}} />
            <div className={style.album}>
                <div>
                    <h2>{info?.title}</h2>
                    <div>
                        <img src={info?.art} alt={`Cover for ${info?.title}`}/>
                        {renderSongList()}
                    </div> 
                </div>
            </div>
        </div>
    );
}

export default Album;