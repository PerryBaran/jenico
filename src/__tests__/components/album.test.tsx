import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import reactRouterDom from "react-router-dom";
import "@testing-library/jest-dom";
import Album from "../../components/routes/album/Album";
import mockData from "../../media/songInfo";

const ALBUM_INDEX = 1;

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("Album", () => {
  const validProps = {
    data: mockData,
    playing: false,
    handlePlaying: jest.fn(),
    selectedSong: mockData[ALBUM_INDEX].songs[0].name,
    handleSelectSong: jest.fn(),
  };

  afterEach(cleanup);

  beforeEach(() => {
    jest
      .spyOn(reactRouterDom, "useParams")
      .mockReturnValue({ album: validProps.data[ALBUM_INDEX].title });
  });

  test("snapshot", () => {
    const { asFragment } = render(
      <Album
        data={validProps.data}
        playing={validProps.playing}
        handlePlaying={validProps.handlePlaying}
        selectedSong={validProps.selectedSong}
        handleSelectSong={validProps.handleSelectSong}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("content", () => {
    render(
      <Album
        data={validProps.data}
        playing={validProps.playing}
        handlePlaying={validProps.handlePlaying}
        selectedSong={validProps.selectedSong}
        handleSelectSong={validProps.handleSelectSong}
      />
    );
    const album = validProps.data[ALBUM_INDEX];

    expect(screen.getByText(album.title)).toHaveAttribute(
      "href",
      album.hyperlink
    );
    expect(screen.getByAltText(`Cover for ${album.title}`)).toHaveAttribute(
      "src",
      album.art
    );
    expect(screen.getByAltText("play/pause album")).toBeInstanceOf(
      HTMLImageElement
    );
    album.songs.forEach((song) => {
      expect(screen.getByText(song.name)).toBeInstanceOf(HTMLButtonElement);
    });
  });

  test("play album with selected Song not in album", () => {
    render(
      <Album
        data={validProps.data}
        playing={validProps.playing}
        handlePlaying={validProps.handlePlaying}
        selectedSong=""
        handleSelectSong={validProps.handleSelectSong}
      />
    );
    const playPause = screen.getByAltText("play/pause album");

    expect(validProps.handlePlaying).toBeCalledTimes(0);
    expect(validProps.handleSelectSong).toBeCalledTimes(0);
    fireEvent.click(playPause);
    expect(validProps.handlePlaying).lastCalledWith(true);
    expect(validProps.handleSelectSong).toBeCalledWith(
      validProps.data[ALBUM_INDEX].songs[0].name
    );
  });

  test("play album with selected Song in album", () => {
    render(
      <Album
        data={validProps.data}
        playing={validProps.playing}
        handlePlaying={validProps.handlePlaying}
        selectedSong={validProps.selectedSong}
        handleSelectSong={validProps.handleSelectSong}
      />
    );
    const playPause = screen.getByAltText("play/pause album");

    expect(validProps.handlePlaying).toBeCalledTimes(0);
    expect(validProps.handleSelectSong).toBeCalledTimes(0);
    fireEvent.click(playPause);
    expect(validProps.handlePlaying).toBeCalledTimes(1);
    expect(validProps.handleSelectSong).toBeCalledTimes(0);
  });

  test("play song from album", () => {
    render(
      <Album
        data={validProps.data}
        playing={validProps.playing}
        handlePlaying={validProps.handlePlaying}
        selectedSong={validProps.selectedSong}
        handleSelectSong={validProps.handleSelectSong}
      />
    );
    const songName = validProps.data[ALBUM_INDEX].songs[1].name;
    const song = screen.getByText(songName);

    expect(validProps.handlePlaying).toBeCalledTimes(0);
    expect(validProps.handleSelectSong).toBeCalledTimes(0);
    fireEvent.click(song);
    expect(validProps.handlePlaying).toBeCalledWith(true);
    expect(validProps.handleSelectSong).toBeCalledWith(songName);
  });
});
