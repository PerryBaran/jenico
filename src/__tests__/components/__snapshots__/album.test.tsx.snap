// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album snapshot 1`] = `
<DocumentFragment>
  <div
    class="albumBackground"
    style="background-image: url(DreamingOfDetunedLove.png);"
  />
  <div
    class="albumContainer"
  >
    <div>
      <h2>
        <a
          href="https://distrokid.com/hyperfollow/jenico/dreaming-of-detuned-love"
          rel="noreferrer"
          target="_blank"
        >
          Dreaming of Detuned Love
        </a>
      </h2>
      <div
        class="album"
      >
        <button>
          <img
            alt="Cover for Dreaming of Detuned Love"
            src="DreamingOfDetunedLove.png"
          />
          <img
            alt="play/pause album"
            src="playCircle.png"
          />
        </button>
        <ul>
          <li
            class="playing"
          >
            <button>
              Slumber
            </button>
          </li>
          <li>
            <button>
              Your Careless Embrace
            </button>
          </li>
          <li>
            <button>
              Restless Thoughts
            </button>
          </li>
          <li>
            <button>
              Detuned Love
            </button>
          </li>
          <li>
            <button>
              Falling into the Void
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
